File: tests/bench_switch.cc
/* //////////////////////////////////////////////////////////////////////////////////////
 * imports
 */
#include <chrono>
#include <cstdio>

#include "base/fiber.h"
#include "base/iomanager.h"
#include "base/laneGo.h"
#include "base/waitGroup.h"

/* //////////////////////////////////////////////////////////////////////////////////////
 * globals
 */
const int COUNT = 10000000;

/* //////////////////////////////////////////////////////////////////////////////////////
 * implementaiton
 */


void switchtask(int count) {

    int i = 0;
    while (true) {

        i++;
        if (i > count) {
            break;
        }
        lane::Fiber::YieldToReady();
    }
}

void TestMain() {
    // get coroutine count
    auto cocount = 1000;

    // init duration
    auto start = std::chrono::high_resolution_clock::now();
    auto wg = new lane::WaitGroup;
    // create coroutine task
    auto i = 0;
    while (true) {

        i++;
        if (i > cocount - 1) {
            break;
        }
        wg->add(1);
        co[=]() {
            switchtask(COUNT / cocount);
            wg->done();
        };
    }
    // computing time
    switchtask(COUNT / cocount);
    wg->wait();

    std::chrono::duration<double> elapsed =
        std::chrono::high_resolution_clock::now() - start;
    // LANE_LOG_INFO(LANE_LOG_ROOT()) << " Test spend " << elapsed.count() << "
    // s";

    // trace

    info() << "switch[" << cocount << "]: lane: " << COUNT << " switches in "
           << elapsed.count() * 1000 << " ms, "
           << int((COUNT / elapsed.count()))
           << " switches per "
              "second";
}

/* //////////////////////////////////////////////////////////////////////////////////////
 * main
 */
int main() {

    // single cpu
    lane::IOManager iom(1, "main", false);
    iom.addTask(TestMain);
}

File: go/bench_switch/bench_switch.go
/* //////////////////////////////////////////////////////////////////////////////////////
 * imports
 */
package main

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"sync"
	"time"
)

/* //////////////////////////////////////////////////////////////////////////////////////
 * globals
 */
const COUNT = 10000000 * 5

/* //////////////////////////////////////////////////////////////////////////////////////
 * implementaiton
 */
var wg sync.WaitGroup

func switchtask(count int) {
	defer wg.Done()
	var i = 0
	for {

		i++
		if i > count {
			break
		}

		runtime.Gosched()
	}
}

/* //////////////////////////////////////////////////////////////////////////////////////
 * main
 */
func main() {

	// single cpu
	runtime.GOMAXPROCS(1)

	// get coroutine count
	cocount := 1000
	if len(os.Args) > 1 {
		cocount, _ = strconv.Atoi(os.Args[1])
	}

	// init duration
	var duration = time.Now().UnixNano()

	// create coroutine task
	var i = 0
	for {

		i++
		if i > cocount-1 {
			break
		}
		wg.Add(1)
		go switchtask(COUNT / cocount)
	}

	// in main goroutine
	switchtask(COUNT / cocount)
	wg.Wait()
	// computing time
	duration = (time.Now().UnixNano() - duration) / 1000000

	// trace
	fmt.Printf("switch[%d]: go: %d switches in %d ms, %d switches per second\n", cocount, COUNT, duration, (1000*COUNT)/duration)
}
